### PR TITLE
Add diagnostics

### PR DESCRIPTION
**Pull request by junpenglao**
Add basic diagnostics
- computation of r-hat
- computation of effective sample size (not based on rank, but easy to add)
- add test to validate sampler accuracy base on monte carlo CLT.

It handles shape flexibly as chain dimension and sample dimension is arbitrary. For example, arviz requires the input array to be [chain, sample, dim0], here we can do [chain, sample, dim0, dim1, ...] or [sample, chain, dim0, dim1, ...], which works both for vmap and pmap output (or even [sample, dim0, dim1, dim2, ..., chain]) 
The diagnostics can be jitted so we can potentially put it in a larger sampling/tuning routine.

### CODE REVIEW

**Comment by twiecki** [comment]
Why have those in JAX and not just use them from arviz?

**Comment by junpenglao** [comment]
I discussed with Remi and the decision is not to add additional dependencies.

The Jax implementation have additional advantage that the shape handling is more flexible: chain dimension and sample dimension is arbitrary (instead arviz requires the input array to be [chain, sample, dimension])

**Comment by twiecki** [comment]
I think adding a dependency is always better (i.e. lower cost) than reimplementing things. And it doesn't even need to be a dependency, this library could just be focused on one thing and do that well (like unix utils): great implementations of samplers implemented in JAX.

**Comment by rlouf** [comment]
I agree that, in general, it is better to have a library that does one thing and does it well. There are several reasons to add some of the diagnostics here:

- By adding a couple hundred lines of codes, we avoid introducing `matplotlib`, `pandas` and `xarray` as dependencies;
- Shape handling is more flexible;
- The measures are jit-able so they can eventually be added to more sophisticated warmup schemes.

We could not do the latter easily by using Arviz.

**Comment by twiecki** [comment]
Regarding the 3 points:
* I think out of those 3 libraries you mentioned, fewer than 1% of users would have an env without `pandas` or `matplotlib`. And if you generate a posterior, you probably will also want to work with it and `xarray` with `arviz` is great for that. I think this will be mostly used as a utility library together with other ones like `numpyro`, `pymc3` etc, and at least `pymc3` also has that dependency, and both of those have `pandas` and `matplotlib`. Finally, none of those are challenging to install.
* That makes sense.
* I would add this functionality here then once you actually want to implement this.

Anyway, I don't think it's a big deal to add this here but wanted to challenge it a bit because feature creep, as we all know, is an easy trap to walk into.

**Comment by rlouf** [comment]
I agree point (1) is not a strong one.

File: blackjax/diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import jax.numpy as jnp
+import numpy as np
+from scipy.fftpack import next_fast_len  # type: ignore
+
+Array = Union[np.ndarray, jnp.DeviceArray]
+
+__all__ = ["potential_scale_reduction", "effective_sample_size"]
+
+
+def potential_scale_reduction(
```
Comment (rlouf) [inline blackjax/diagnostics.py:14]: So if I want to compute the Rhat value in a problem with 2 variables, do I have to use `tree_multimap` to compute the values?

File: blackjax/diagnostics.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    input_array: Array, chain_axis: int = 0, sample_axis: int = 1
+):
+    """Gelman and Rubin (1992)'s potential scale reduction for computing multiple MCMC chain convergence.
```
Comment (rlouf) [inline blackjax/diagnostics.py:17]: Could you add a reference in the docstring?

File: blackjax/diagnostics.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        (between_chain_variance / within_chain_variance + num_samples - 1)
+        / (num_samples)
+    )
+    return rhat_value.squeeze()
+
+
+def effective_sample_size(
+    input_array: Array, chain_axis: int = 0, sample_axis: int = 1
+):
+    """Compute estimate of the effective sample size (ess).
```
Comment (rlouf) [inline blackjax/diagnostics.py:55]: Could you add a reference to a paper/book in which this algorithm is described?

File: tests/test_diagnostics.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+"""Test MCMC diagnoistics."""
```
Comment (rlouf) [inline tests/test_diagnostics.py:1]:
```suggestion
"""Test MCMC diagnostics."""
```

File: tests/test_diagnostics.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        input_list.insert(loc, elem)
+    return input_list
+
+
+@pytest.mark.parametrize("case", test_cases)
+@pytest.mark.parametrize("nchain", [2, 10])
+@pytest.mark.parametrize("event_shape", [(), (3,), (5, 7)])
+def test_rhat_ess(case, nchain, event_shape):
+    rng_key = jax.random.PRNGKey(32)
+    n_samples = 5000
```
Comment (rlouf) [inline tests/test_diagnostics.py:41]: nitpicking: `n_samples` and `nchain` use different naming conventions

Comment (rlouf) [reply tests/test_diagnostics.py:41]: I prefer `n_chains` (or better `num_chains`)

**Review by rlouf — COMMENTED** [review]
This is a superficial review, I focused on the form. I still have to look at the algorithms in detail.

File: blackjax/diagnostics.py (hunk 1 above)
Comment (junpenglao) [reply blackjax/diagnostics.py:14]: Use `tree_map` but yes the function takes arrays as input

File: tests/test_diagnostics.py (hunk 5 above)
Comment (junpenglao) [reply tests/test_diagnostics.py]: Done

**Comment by rlouf** [comment]
Great work ! Everything looks good to me, merging.

**Comment by OriolAbril** [comment]
I saw this today and wanted to clarify one point, not to change the outcome here but to try to avoid confusion in the future.

> arviz requires the input array to be [chain, sample, dim0]

This is not really true, but I do understand where it comes from. The first and most important thing to keep in mind is that ArviZ is built on top of xarray. It is not build on top of numpy nor jax nor pytorch nor any other array library. In xarray dimension order plays no role by design, the only thing that matters are dimension names (as shown for example in https://github.com/arviz-devs/arviz/issues/1693#issuecomment-840681873). **What ArviZ requires is that there is one dimension _named_ `chain` and another _named_ `draw`**.

When working with array data, it is first converted to xarray format and then passed to the actual function. It is in this conversion
step where the limitation is. Currently all converters use that `chain, sample, *shape` convention to _name_ he dimensions when creating the corresponding xarray objects, but he converters can be updated at any time if someone has time for it, and anyone can write custom converters at any point.
